File: tests/testLsf1d.py
#!/usr/bin/env python

#
# LSST Data Management System
# Copyright 2008, 2009, 2010 LSST Corporation.
#
# This product includes software developed by the
# LSST Project (http://www.lsst.org/).
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the LSST License Statement and
# the GNU General Public License along with this program.  If not,
# see <http://www.lsstcorp.org/LegalNotices/>.
#

import unittest

import lsst.utils.tests
import lsst.pex.exceptions
import lsst.meas.astrom.sip as sip


class Lsf1dTestCase(unittest.TestCase):

    def testBadArgs1(self):
        """Check that code fails when order is too low"""

        x = [630, 1000, 1205, 1427]
        y = [.622, 1.109, 1.198, 1.429]
        s = [1, 1, 1, 1]

        order = 0

        with self.assertRaises(lsst.pex.exceptions.Exception):
            sip.LeastSqFitter1dPoly(x, y, s, order)

    def testBadArgs2(self):
        """Check that code fails when not len(x) != len(y)"""

        x = [630, 1000, 1205, 1427]
        y = [.622, 1.109, 1.198]
        s = [1, 1, 1, 1]

        order = 0

        with self.assertRaises(lsst.pex.exceptions.Exception):
            sip.LeastSqFitter1dPoly(x, y, s, order)

    def testBadArgs3(self):
        """Check that code fails when not len(x) != len(s)"""

        x = [630, 1000, 1205, 1427]
        y = [.622, 1.109, 1.198, 1.429]
        s = [1, 1, 1, 1, 1]

        order = 0

        with self.assertRaises(lsst.pex.exceptions.Exception):
            sip.LeastSqFitter1dPoly(x, y, s, order)

    def testBadArgs4(self):
        """Check that code fails when not order > number data points"""

        x = [630, 1000, 1205, 1427]
        y = [.622, 1.109, 1.198, 1.429]
        s = [1, 1, 1, 1]

        order = 5

        with self.assertRaises(lsst.pex.exceptions.Exception):
            sip.LeastSqFitter1dPoly(x, y, s, order)

    def testConst1(self):
        """Check that code fits a dc offset correctly (1)"""

        x = [630, 1000, 1205, 1427]
        y = [.622, 1.109, 1.198, 1.429]
        s = [1, 1, 1, 1]

        order = 1

        lsf = sip.LeastSqFitter1dPoly(x, y, s, order)

        for i in range(1, len(x)):
            self.assertAlmostEqual(lsf.valueAt(x[0]), lsf.valueAt(x[i]), 4)
            # print x[i], y[i], lsf.valueAt(x[i])

    def testCompareToCpp1(self):
        """Confirm that I get the same behaviour in Python as I
        do in the C++ test fitLinear3
        """

        # This test arises because I was having trouble using Eigen's
        # svd inversion. fitLinear3 in C++ worked fine, but this test
        # behaved differently, and I tracked the result down to
        # the svd routines.

        x = [689.301136505, 1112.8573687, 1386.67168477]
        y = [0.66911456573, 1.1147439759, 1.39597284177]
        s = [1, 1, 1]

        order = 2
        print
        lsf = sip.LeastSqFitter1dPoly(x, y, s, order)

        self.assertAlmostEqual(lsf.valueAt(x[0]), 0.670187891961, 4)
        self.assertAlmostEqual(lsf.valueAt(x[1]), 1.11201034929, 4)
        self.assertAlmostEqual(lsf.valueAt(x[2]), 1.39763314215, 4)

    def testCompareToCpp2(self):
        """Confirm that I get the same behaviour in Python as I do in the C++ test fitLinear3
        """

        # This test arises because I was having trouble using Eigen's
        # svd inversion. fitLinear3 in C++ worked fine, but this test
        # behaved differently, and I tracked the result down to
        # the svd routines.

        x = [628.857680996, 995.008255088, 1203.39412154, 1425.1404727]
        y = [0.616728229875, 1.01887634344, 1.19679830465, 1.42873084062]
        s = [1, 1, 1, 1]

        order = 3
        print
        lsf = sip.LeastSqFitter1dPoly(x, y, s, order)

        for i in range(len(x)):
            f = lsf.valueAt(x[i])
            print "%.1f %.3f %.3f" % (x[i], y[i], f)

        self.assertAlmostEqual(lsf.valueAt(x[0]), y[0], 1)
        self.assertAlmostEqual(lsf.valueAt(x[1]), y[1], 1)
        self.assertAlmostEqual(lsf.valueAt(x[2]), y[2], 1)


class MemoryTester(lsst.utils.tests.MemoryTestCase):
    pass


def setup_module(module):
    lsst.utils.tests.init()


if __name__ == "__main__":
    lsst.utils.tests.init()
    unittest.main()

File: tests/test_lsf1d.cc
/*
 * LSST Data Management System
 * Copyright 2008, 2009, 2010 LSST Corporation.
 *
 * This product includes software developed by the
 * LSST Project (http://www.lsst.org/).
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the LSST License Statement and
 * the GNU General Public License along with this program.  If not,
 * see <http://www.lsstcorp.org/LegalNotices/>.
 */

#define BOOST_TEST_DYN_LINK
#define BOOST_TEST_MODULE testlsf1d

// The boost unit test header
#include "boost/test/unit_test.hpp"

using namespace std;

#include <vector>
#include <cmath>
#include <cstdio>
#include <cassert>
#include <memory>
#include "Eigen/Core"
#include "lsst/afw/math/FunctionLibrary.h"

#include "lsst/meas/astrom/sip/LeastSqFitter1d.h"

namespace sip = lsst::meas::astrom::sip;
namespace math = lsst::afw::math;

BOOST_AUTO_TEST_CASE(fitLine) {
    vector<double> x;
    vector<double> y;
    vector<double> s;

    for (int i = 0; i < 7; ++i) {
        x.push_back((double)i);
        y.push_back((double)3 + 2 * i);
        s.push_back((double)1);
    }

    int order = 2;
    sip::LeastSqFitter1d<math::PolynomialFunction1<double> > lsf(x, y, s, order);

    Eigen::VectorXd par = lsf.getParams();

    BOOST_CHECK(par.size() == order);
    BOOST_CHECK_CLOSE(par(0), 3., .001);
    BOOST_CHECK_CLOSE(par(1), 2., .001);
}

BOOST_AUTO_TEST_CASE(fitQuadratic) {
    vector<double> x;
    vector<double> y;
    vector<double> s;

    for (int i = 0; i < 7; ++i) {
        x.push_back((double)i);
        y.push_back((double)4 + i * (3 + i * 2));
        s.push_back((double)1);
    }

    int order = 3;
    sip::LeastSqFitter1d<math::PolynomialFunction1<double> > lsf(x, y, s, order);

    math::PolynomialFunction1<double> f = lsf.getBestFitFunction();

    BOOST_CHECK((int)f.getNParameters() == order);
    BOOST_CHECK_CLOSE(f.getParameter(0), 4., .001);
    BOOST_CHECK_CLOSE(f.getParameter(1), 3., .001);
    BOOST_CHECK_CLOSE(f.getParameter(2), 2., .001);

    BOOST_CHECK_CLOSE(lsf.valueAt(1.), 9., .001);
}

// Check a more extreme case where there's a big difference between the x and y values
BOOST_AUTO_TEST_CASE(fitLinear2) {
    vector<double> x;
    vector<double> y;
    vector<double> s;

    x.push_back(631.920343622);
    x.push_back(1000.02009955);
    x.push_back(1205.63487689);
    x.push_back(1427.86231185);

    y.push_back(0.622249643993);
    y.push_back(1.01887634344);
    y.push_back(1.1982950985);
    y.push_back(1.4294170431);

    s.push_back(1);
    s.push_back(1);
    s.push_back(1);
    s.push_back(1);

    int order = 2;
    sip::LeastSqFitter1d<math::PolynomialFunction1<double> > lsf(x, y, s, order);
    Eigen::VectorXd par = lsf.getParams();
    assert(par[0] == par[0]);  // stop compiler whining about par not being used
}

BOOST_AUTO_TEST_CASE(fitLinear3) {
    vector<double> x;
    vector<double> y;
    vector<double> s;

    x.push_back(689.301136505);
    x.push_back(1112.8573687);
    x.push_back(1386.67168477);

    y.push_back(0.66911456573);
    y.push_back(1.1147439759);
    y.push_back(1.39597284177);

    s.push_back(1);
    s.push_back(1);
    s.push_back(1);

    int order = 2;
    sip::LeastSqFitter1d<math::PolynomialFunction1<double> > lsf(x, y, s, order);
    Eigen::VectorXd par = lsf.getParams();

    BOOST_CHECK_CLOSE(par(0), -0.0488399, .001);
    BOOST_CHECK_CLOSE(par(1), 0.00104313, .001);
}

BOOST_AUTO_TEST_CASE(fitQuadratic2) {
    vector<double> x;
    vector<double> y;
    vector<double> s;

    x.push_back(628.857680996);
    x.push_back(995.008255088);
    x.push_back(1203.39412154);
    x.push_back(1425.1404727);

    y.push_back(0.61672822987);
    y.push_back(1.01887634344);
    y.push_back(1.19679830465);
    y.push_back(1.42873084062);

    s.push_back(1);
    s.push_back(1);
    s.push_back(1);
    s.push_back(1);

    int order = 3;
    sip::LeastSqFitter1d<math::PolynomialFunction1<double> > lsf(x, y, s, order);
    Eigen::VectorXd par = lsf.getParams();
    assert(par[0] == par[0]);  // stop compiler whining about par not being used

    for (unsigned int i = 0; i != x.size(); ++i) {
        // printf("%.3f %.3f %.3f \n", x[i], y[i], lsf.valueAt(x[i]));
        BOOST_CHECK_CLOSE(y[i], lsf.valueAt(x[i]), 10);
    }
}

BOOST_AUTO_TEST_CASE(errorbars) {
    vector<double> x;
    vector<double> y;
    vector<double> s;

    x.push_back(1);
    x.push_back(2);
    x.push_back(3);

    y.push_back(1);
    y.push_back(2);
    y.push_back(3);

    s.push_back(1);
    s.push_back(1);
    s.push_back(1);

    int order = 2;
    sip::LeastSqFitter1d<math::PolynomialFunction1<double> > lsf(x, y, s, order);
    Eigen::VectorXd par = lsf.getParams();
    Eigen::VectorXd err = lsf.getErrors();

    // Boost doesn't do well at checking for zero
    BOOST_CHECK_CLOSE(par[0] + 1, 1., 1e-6);
    BOOST_CHECK_CLOSE(par[1], 1., 1e-6);

    // Calculated by hand
    BOOST_CHECK_CLOSE(err[0], 1.52752523165195, 1e-6);
    BOOST_CHECK_CLOSE(err[1], 0.7071067811865481, 1e-6);
}
